netlify.toml

```diff
@@ -6,7 +6,7 @@
 
 [context.deploy-preview.build]
   base = "docs"
-  ignore = "git diff --quiet $CACHED_COMMIT_REF $COMMIT_REF ."
+  ignore = "git diff --quiet HEAD^ HEAD ."
```

Comment:
Can you explain what you're trying to accomplish here?  I think what you're trying to do is ave unnecessary builds but I'm not sure this will accommodate skipping builds on multi commit PRs or across the, e.g., correct side of merge branches, etc.

Reply:
This `git diff` runs inside the base directory (docs) so it's meant to check to see if the current PR adds any commits in that directory. Can you explain what "the correct side of merge branches" means?

Reply:
I think this will work if you're comfortable working under the assumption that:

- PRs will have only a single commit to their base branch
- PRs will never have merge commits in them themselves

Don't take this as a perfect claim but, roughly, I'd claim that:

The `HEAD^` syntax refers to the parent of the current commit (`HEAD`).  It is effectively the same as `HEAD~` but used to refer to a merge commits' members (note this does not mean base branch). The caret is the merge signifier and it specifically refers to the first merge member in a merge commit (but it's worth noting that merge commits can have "n" members; there is such thing as `^5` or `HEAD^^^^^` - which are the same). 

I think that [this page](https://medium.com/@gabicle/git-ancestry-references-explained-bd3a84a0b821) probably does a better job explaining better than the official git docs since it has different examples but the official git docs [have a section called "Ancestry References](https://git-scm.com/book/en/v2/Git-Tools-Revision-Selection) (no anchor link, sorry!)

I think what you want for deploy previews is a diff against the target branch.  I'd claim that's an environment variable that Netlify owes you, in the same way that CircleCI and other providers offer it: i think it's in their docs [as `$HEAD`](https://docs.netlify.com/configure-builds/environment-variables/#git-metadata), but don't conflate that with `HEAD`!

Anyhow, i'm not fully at my computer right now but hope this helps!

Reply:
Oh i should add: i don't know if those stated assumptions i started with are okay to expect of a PR, and i'm challenging (hopefully nicely!) whether they're what you desire. 😉 

i could be misunderstanding though!

Reply:
I stole the line from this blog post: https://www.netlify.com/blog/2020/04/27/ignore-unnecessary-builds-to-optimize-your-build-times/

Reply:
even though that blog post suggests it, i don't think that invalidates my claim and i think my link supports my claim.

the worst case scenario here is that we get odd behavior or missed builds from our documentation builds.  both of those seem pretty important to avoid though, so we might want to validate that it works as expected in common workflows!

Reply:
curious: was what was here before not working?

Reply:
Yeah, the previous behavior was causing builds to happen unnecessarily, probably because `$CACHED_COMMIT_REF` wasn't set to anything on netlify's end. In any case, netlify would let the build proceed even though the PR contained no commits affecting files in the base directory (`docs`).

Thanks for the background about the `HEAD` syntax. I didn't mean to send that blog post link to invalidate your claim! I didn't see your subsequent messages before I posted that comment, I just meant for it to follow my previous comment.

I guess what I'm trying to express in this `ignore` statement is: "does the current commit change anything in the docs compared to the latest commit in that directory?" I can see that the way that I've written it now has some limitations and maybe we're better off using the previous strategy, and accepting that sometimes non-docs PRs will trigger a deploy preview. Or maybe there's something we can add to the previous `ignore` command that will account for netlify not having a `$CACHED_COMMIT_REF` handy for some reason.

Reply:
@trevorblades Sorry, I forgot about this as I went into my holiday!


One possible alternative to using `$CACHED_COMMIT_REF` would be to employ [`git merge-base --fork-point`](https://git-scm.com/docs/git-merge-base#Documentation/git-merge-base.txt---fork-point). I'm not actually sure what that would do on a flat commit history.

Looking more closely at [the Netlify docs](https://docs.netlify.com/configure-builds/environment-variables/#git-metadata) though, perhaps the environment variable that you want is actually literally called `HEAD`  (as in, **you would use it by using `$HEAD`**?).  That's not very intuitively named, imho because HEAD most definitely **not** the branch you're merging into but 🤷  

So that means that perhaps you'd use (confusingly) `git diff --quiet $HEAD HEAD .` as the option? 🤔   It does seem to also indicate that `$CACHED_COMMIT_REF` would be the last successful build on this branch, which makes sense why it wasn't reliably present!

Reply:
No worries at all. I got a notification for this while I was on vacation and I forgot to follow up too! 😖

I just tried out using `git diff --quiet $HEAD HEAD`, but it produced the following logs:

```
5:20:56 PM: Detected ignore command in Netlify configuration file. Proceeding with the specified command: 'git diff --quiet $HEAD HEAD .'
5:20:56 PM: fatal: ambiguous argument 'tb/docs-publish': unknown revision or path not in the working tree.
5:20:56 PM: Use '--' to separate paths from revisions, like this:
5:20:56 PM: 'git <command> [<revision>...] -- [<file>...]'
```

This seems like the `HEAD` environment variables = `tb/docs-publish` (the name of this branch). I agree that what we want is a reference to the branch that the current PR is merging into.

What if we just did `git diff --quiet main HEAD .`? I just tried it out and didn't get any complaints from Netlify. Do you think this approach would be suitable?